### ISSUE TITLE
Project Validation

### ISSUE THREAD

**Issue by arichiv-usds** [Code-dot-mil/code.mil#186]
We should validate the JSON before pushing any changes that could break or improperly update project info.

**Comment by jgarber623-gov** [comment]
Possibly useful links:

- [Code.gov's online schema validator](https://code.gov/#/policy-guide/docs/compliance/inventory-code/tools/validate-schema)
- [The source code](https://github.com/GSA/code-gov-web/tree/master/src/app/components/tools/schema-validator) for Code.gov's online schema validator (written in TypeScript)

**Comment by arichiv-usds** [comment]
Nice

**Comment by jordangov** [comment]
Yeah... FYI, I've talked to Alvand and they'd like to get the schema validation into the API, but it is not there right now. We may want to wait for that versus doing this.

**Comment by arichiv-usds** [comment]
Maybe I could add that :-)

**Comment by arichiv** [comment]
https://github.com/GSA/code-gov-web/issues/564

**Comment by DanielJDufour** [comment]
FYI: We've updated our validator, which you can find here: https://code.gov/about/compliance/inventory-code/validate-schema

It's built with [json-schema-validator-web-component](https://github.com/GSA/json-schema-validator-web-component) a reusable web component! :-)

**Comment by seakintruth** [comment]
Similar to the Code.json Validator it would be helpful to report broken URL links in the Code.json file.

I put together a google sheets prototype of sorts here:
[https://docs.google.com/spreadsheets/d/13tzylBYFAISHDaN9QOgL-qIj4we0V7X_Jz_pyzKYrxU/edit?usp=sharing](https://docs.google.com/spreadsheets/d/13tzylBYFAISHDaN9QOgL-qIj4we0V7X_Jz_pyzKYrxU/edit?usp=sharing)

It publishes a daily report to here:
[https://docs.google.com/spreadsheets/d/e/2PACX-1vQm1kqCv30grzurClLrYTGb3P-itY07Cbl70Hu-7UBEIAKAiz3s8v1y-VFAp4vLqaH3qQ2FVilRSFHb/pubhtml](https://docs.google.com/spreadsheets/d/e/2PACX-1vQm1kqCv30grzurClLrYTGb3P-itY07Cbl70Hu-7UBEIAKAiz3s8v1y-VFAp4vLqaH3qQ2FVilRSFHb/pubhtml)

This tool attempts to daily import each agency's code.json file as described by the code.gov site and verify pertinent URLs for each agency's code.json file. These verifications are displayed on the UrlStatus worksheet. This report displays the HTTPS response codes when >=300 from a  call to Google App Script UrlFetch() for each URL.	

It works as is, but only validates files that are found from the existing agencies code.json files. I get this form a private sheet's range that's pulling from https://code-api.app.cloud.gov/api/agencies?size=10000&api_key={My Key}

**Comment by jordangov** [comment]
Neat! Have you also posted that to the [Code.gov](https://github.com/GSA/code-gov) repo?

**Comment by seakintruth** [comment]
I just posted this as a comment to [Code.gov Administration Tool for Agency Inventory Upload and Maintenance #29](https://github.com/GSA/code-gov/issues/29)

**Comment by jalbertbowden** [comment]
Curious why Project Validation is only JSON validation? I'm sure it'll fail if JSON doesn't validate, but wondering the emphasis behind this?
W3C has a linkrot validator, amongst their other validators, that you could use here too:
https://validator.w3.org/checklink

**Comment by jordangov** [comment]
Happy to include that as well...

**Comment by afomi** [comment]
Is there more work to be done for this issue?

It looks like JSON is being validated.  If there's more to be validated, perhaps it can be clarified in a new issue.

**Comment by jordangov** [comment]
The format of the JSON is validated, but not the structure that Code.gov really wants, or the content of the information (like that URLs really are URLs, or that there is a POC).
